Python/bytecodes.c

```diff
@@ -283,44 +282,48 @@ dummy_func(
         }
 
         pure inst(PUSH_NULL, (-- res)) {
-            res = NULL;
+            res = Py_STACKREF_NULL;
         }
 
         macro(END_FOR) = POP_TOP;
 
         tier1 inst(INSTRUMENTED_END_FOR, (receiver, value -- receiver)) {
             /* Need to create a fake StopIteration error here,
              * to conform to PEP 380 */
-            if (PyGen_Check(receiver)) {
-                if (monitor_stop_iteration(tstate, frame, this_instr, value)) {
+            if (PyGen_Check(PyStackRef_AsPyObjectBorrow(receiver))) {
+                if (monitor_stop_iteration(tstate, frame, this_instr, PyStackRef_AsPyObjectBorrow(value))) {
                     ERROR_NO_POP();
                 }
             }
             DECREF_INPUTS();
         }
 
         pure inst(END_SEND, (receiver, value -- value)) {
-            Py_DECREF(receiver);
+            (void)receiver;
+            PyStackRef_CLOSE(receiver);
         }
 
         tier1 inst(INSTRUMENTED_END_SEND, (receiver, value -- value)) {
-            if (PyGen_Check(receiver) || PyCoro_CheckExact(receiver)) {
-                if (monitor_stop_iteration(tstate, frame, this_instr, value)) {
+            PyObject *receiver_o = PyStackRef_AsPyObjectBorrow(receiver);
+            if (PyGen_Check(receiver_o) || PyCoro_CheckExact(receiver_o)) {
+                if (monitor_stop_iteration(tstate, frame, this_instr, PyStackRef_AsPyObjectBorrow(value))) {
                     ERROR_NO_POP();
                 }
             }
-            Py_DECREF(receiver);
+            PyStackRef_CLOSE(receiver);
         }
 
         inst(UNARY_NEGATIVE, (value -- res)) {
-            res = PyNumber_Negative(value);
+            PyObject *res_o = PyNumber_Negative(PyStackRef_AsPyObjectBorrow(value));
             DECREF_INPUTS();
-            ERROR_IF(res == NULL, error);
+            ERROR_IF(res_o == NULL, error);
+            res = PyStackRef_FromPyObjectSteal(res_o);
         }
 
         pure inst(UNARY_NOT, (value -- res)) {
-            assert(PyBool_Check(value));
-            res = Py_IsFalse(value) ? Py_True : Py_False;
+            assert(PyBool_Check(PyStackRef_AsPyObjectBorrow(value)));
+            res = PyStackRef_FromPyObjectSteal(Py_IsFalse(PyStackRef_AsPyObjectBorrow(value))
```

Comment:
Maybe add PyRef_IsFalse/True/None to save all the borrow/steal stuff.
```suggestion
            res = PyRef_IsFalse(value) ? PyRef_True : PyRef_False;
```

Reply:
Thanks. I'm gonna do this in a separate PR after this one. I recall during the sync @colesbury had some concerns about  this, but I don't recall what? Or maybe I'm misremembering things.

Reply:
@colesbury was concerned that untagging in comparisons would be expensive.
In other words, implement this:
```C
bool PyRef_IsFalse(PyStackRef ref) {
    ref.bits == FALSE_BITS;
}
```
not this:
```C
bool PyRef_IsFalse(PyStackRef ref) {
    UNTAG(ref) == Py_False;
}
```

Reply:
Done.